### PR TITLE
SWIFT-1041 Implement support for setup and teardown code via an extension to Application

### PR DESCRIPTION
**Pull request by ncarbon**
This just adds the setup and cleanup methods.

### CODE REVIEW

**Review by patrickfreed — COMMENTED** [review]
Code changes look good, but can we add some tests for the various behaviors?

File: Tests/MongoDBVaporTests/MongoDBVaporTests.swift [hunk 1]
```diff
@@ -1,6 +1,20 @@
...
+        var env = try Environment.detect()
+        try LoggingSystem.bootstrap(from: &env)
+        let app = Application(env)
+
+        try app.configureMongoDB()
+        expect(app.mongoClient).toNot(beNil())
+        expect(try app.mongoClient.listDatabases().wait()).toNot(throwError())
+
+        try app.cleanupMongoDB()
+        expect(try app.mongoClient.listDatabases().wait()).to(throwError())
```
Comment (ncarbon) [inline Tests/MongoDBVaporTests/MongoDBVaporTests.swift]: Should we test that a specific error is thrown when trying to use a client that has already been closed? Like a `ChannelError` (NIO)?

Comment (patrickfreed) [reply Tests/MongoDBVaporTests/MongoDBVaporTests.swift]: I think it's fine that we just test that some error occurs here. I don't think its necessarily that important that we test which error is thrown (at least not in this library). We could add such a test in the driver if need be.

File: Sources/MongoDBVapor/MongoDBVapor.swift [hunk 2]
```diff
@@ -1 +1,49 @@
...
+    }
+
+    private struct MongoClientKey: StorageKey {
+        // swiftlint:disable nesting
+        typealias Value = MongoClient
+    }
+
+    /**
+     * Configures a global `MongoClient` with the specified options. The client will use the application's
+     * `EventLoopGroup` for executing operations.
```
Comment (patrickfreed) [inline Sources/MongoDBVapor/MongoDBVapor.swift]: can we add that that the client may be accessed via the `mongoClient` computed property on the `Application`? Also perhaps a code sample would be nice. Check [here](https://developer.apple.com/library/archive/documentation/Xcode/Reference/xcode_markup_formatting_ref/CodeBlocks.html#//apple_ref/doc/uid/TP40016497-CH12-SW1) for the syntax / format. I think four backticks (\`\`\`\`) seems preferable to the indentation based format.

**Review by patrickfreed — APPROVED** [review]
LGTM!

**Comment by kmahar** [comment]
@0xTim I can't seem to request a code review from you but I just invited you to have "read" permission for this repo which might enable that? In any case, this is the PR I mentioned 🙂

File: Sources/MongoDBVapor/MongoDBVapor.swift [hunk 3]
```diff
@@ -1 +1,65 @@
...
+     *  app.shutdown()
+     * ````
+     */
+    public func cleanupMongoDB() {
+        do {
+            try self.mongoClient.syncClose()
+        } catch {
+            self.logger.error("Failed to shut down MongoDB client: \(error)")
+        }
+        cleanupMongoSwift()
```
Comment (kmahar) [inline Sources/MongoDBVapor/MongoDBVapor.swift]: @0xTim a question: is having multiple `Application`s running at once (like via the same `Run` executable) something Vapor users would ever do? Or is it fairly safe to assume a 1:1 relationship?

This `cleanupMongoSwift` method is an unfortunate artifact of our wrapping the C driver which has a global cleanup method `mongoc_cleanup` we're wrapping here: http://mongoc.org/libmongoc/1.16.2/mongoc_cleanup.html 

Basically it frees some global state and after its called the driver is no longer usable. So if people use multiple `Application`s we could run into an issue here if only one has been shut down and others are still trying to use the driver.

Comment (0xTim) [reply Sources/MongoDBVapor/MongoDBVapor.swift]: Anyone running multiple `Application`s in the same process is definitely an edge case. I'm not sure Vapor would even allow it. So you can probably safely assume it's fine for now

**Comment by 0xTim** [comment]
Mostly looks good to me! Couple of tweaks I'd suggest:
* Definitely add an extension for accessing the MongoDB client via `Request` - you can then ensure that any MongoDB requests are made on the right event loop to avoid having to hop back onto the request event loop
* The Vapor idiom would be to have something like `application.mongoClient.configuration` for configuring the client rather than a separate function on `application`. Effectively that namespaces it to the view of the end-user but that's definitely a stylistic change to match the rest of the ecosystem rather than anything necessary

File: Sources/MongoDBVapor/MongoDBVapor.swift (hunk 3 above)
Comment (kmahar) [reply Sources/MongoDBVapor/MongoDBVapor.swift]: thanks for that context! writing some more tests I did realize I think this actually will cause a problem in test code though - if users create separate `Application`s within different test cases (as I am now doing) and shut them down then we run into the same issue with multiple calls.
so I think we will still have to require users to call this themselves when they are done rather than handling it for them here. oh well.

**Review by kmahar — COMMENTED** [review]
Thanks very much for the input! 

I've done some refactoring to namespace the configuration/cleanup code within `app.mongoDB` (heavily inspired by the Redis provider library...).  Also went ahead and added an extension to `Request` with the same namespacing to access a client on the right event loop.

**Comment by kmahar** [comment]
oh hmm. I don't seem to actually be able to push to Nataly's branch. she was doing a rotation on our team and started this work but I'm picking up the rest. let me see if I can figure this out....

**Comment by kmahar** [comment]
nvm, figured it out 🙂

**Review by 0xTim — APPROVED** [review]
LGTM! Just out of curiosity, any particular reason for using Nimble?

File: Sources/MongoDBVapor/MongoDBVapor.swift (hunk 3 above)
Comment (0xTim) [reply Sources/MongoDBVapor/MongoDBVapor.swift]: Do you mean each test having a separate application? That should be supported and fine unless you have any global static config 🤔 But adding in the docs to tell them to shut it down should be fine

**Review by patrickfreed — APPROVED** [review]

File: Sources/MongoDBVapor/MongoDBVapor.swift (hunk 3 above)
Comment (kmahar) [reply Sources/MongoDBVapor/MongoDBVapor.swift]: yeah there is some global stuff the C driver is doing unfortunately

**Comment by kmahar** [comment]
@0xTim thanks! re Nimble not in particular, we just like the style of writing expectations so we switched to it in the driver way back and have stuck with it since 🙂 you can add custom matcher functions, so it's been really convenient there to support some more complex assertions we often make.
